File: backend/src/app/schemas/tracking.py
"""Schemas for tracking (user list entries)."""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from src.app.models.enums import WatchStatus


class ListEntryCreate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    media_id: UUID
    status: WatchStatus
    progress: int = 0
    score: Optional[float] = None
    notes: Optional[str] = None
    is_private: bool = False
    repeat_count: int = 0
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class ListEntryUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    status: Optional[WatchStatus] = None
    progress: Optional[int] = None
    score: Optional[float] = None
    notes: Optional[str] = None
    is_private: Optional[bool] = None
    repeat_count: Optional[int] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class ListEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    media_id: UUID
    status: WatchStatus
    progress: int
    score: Optional[float] = None
    notes: Optional[str] = None
    is_private: bool
    repeat_count: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class UserListResponse(BaseModel):
    items: list[ListEntryResponse]
    total: int
    limit: int
    offset: int


class ListEntryHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    entry_id: UUID
    user_id: UUID
    media_id: UUID
    event_type: str
    old_status: Optional[WatchStatus] = None
    new_status: Optional[WatchStatus] = None
    old_progress: Optional[int] = None
    new_progress: Optional[int] = None
    old_score: Optional[float] = None
    new_score: Optional[float] = None
    note: Optional[str] = None
    created_at: datetime


class UserListHistoryResponse(BaseModel):
    items: list[ListEntryHistoryResponse]
    total: int
    limit: int


class CustomListCreate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    description: Optional[str] = None
    is_public: bool = False
    cover_image: Optional[str] = None
    sort_order: int = 0


class CustomListResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    name: str
    description: Optional[str] = None
    is_public: bool
    cover_image: Optional[str] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime


class CustomListEntryUpsert(BaseModel):
    media_id: UUID
    sort_order: int = 0
    note: Optional[str] = None


class CustomListEntriesReplaceRequest(BaseModel):
    entries: list[CustomListEntryUpsert]


class CustomListEntriesReplaceResponse(BaseModel):
    list_id: UUID
    total_entries: int

File: frontend/src/types/tracking.ts
export type WatchStatus =
  | 'watching'
  | 'reading'
  | 'completed'
  | 'paused'
  | 'dropped'
  | 'plan_to_watch'
  | 'plan_to_read'
  | 'rewatching'
  | 'rereading'

export interface ListEntry {
  id: string
  media_id: string
  status: WatchStatus
  progress: number
  score?: number | null
  notes?: string | null
  title?: string
  cover_image_medium?: string | null
}

export interface UserListResponse {
  items: ListEntry[]
  total: number
  limit: number
  offset: number
}

export interface ListEntryCreate {
  media_id: string
  status: WatchStatus
  progress?: number
  score?: number
  notes?: string
}

export interface ListEntryUpdate {
  status?: WatchStatus
  progress?: number
  score?: number | null
  notes?: string
}

export interface CustomList {
  id: string
  name: string
  description?: string | null
  is_public: boolean
  cover_image?: string | null
  sort_order: number
}

export interface CustomListCreate {
  name: string
  description?: string
  is_public?: boolean
  cover_image?: string
  sort_order?: number
}

export interface CustomListEntriesReplaceRequest {
  entries: Array<{
    media_id: string
    sort_order?: number
    note?: string
  }>
}

// -- Phase 5.3: History & Statistics --

export interface ListEntryHistoryItem {
  id: string
  entry_id: string
  media_id: string
  event_type: string
  old_status?: WatchStatus | null
  new_status?: WatchStatus | null
  old_progress?: number | null
  new_progress?: number | null
  old_score?: number | null
  new_score?: number | null
  note?: string | null
  created_at: string
}

export interface UserListHistoryResponse {
  items: ListEntryHistoryItem[]
  total: number
  limit: number
}

export interface ListStats {
  total: number
  watching: number
  reading: number
  completed: number
  paused: number
  dropped: number
  plan_to_watch: number
  plan_to_read: number
  rewatching: number
  rereading: number
}
